app/models/user.rb

```diff
@@ -206,6 +206,19 @@ def otp_verified?(otp)
     save!(validate: false)
   end
 
+  def lock!
+    transaction do
+      update_attribute(:email, "security+locked-#{handle}@rubygems.org")
```

Comment:
Is there a potential issue here if the bad actor registers 2 user accounts where one uses the generated email here? Say if this occurred:

1. Bad actor registers user A with an email address "security+locked-badactor@rubygems.org"
2. Bad actor registers user B user with handle "badactor"
3. Bad actor does something as user B which leads admin to attempt to lock user
4. Locking user B fails as the email "security+locked-badactor@rubygems.org" is already taken by user A

Reply:
I understand your concern. This locking strategy was blindly copied from original script to make as small change as possible. If I understand well, email is changed to not make possible to recovery password for user.

To fix described scenario, isn't it enough to include also id?

```ruby
update_attribute(:email, "security+locked-#{id}-#{handle}@rubygems.org")
```

Reply:
This is an improvement, although its still predictable, especially if user ids are exposed.

Consider using a random value instead of `id` and separate it with `+` so its clearer the random value is part of the prefix and not part of the `handle`, using the `-` only to separate the `handle`.

```rb
"security+locked+#{SecureRandom.some_method}-#{handle}@rubygems.org"
```

Reply:
User IDs means that it's guaranteed to be unique, but I'm fine with either uid or some random bit.

Reply:
Its not that its unique, its that it can be predicted and reserved by registering as another user.

1. Bad actor registers user A with handle `usera`. User A has id 123456.
2. Bad actor registers user B with email address `security+locked+123456-usera@rubygems.org`
3. `User.find(handle: 'usera').block!` fails as it the generated email address is already taken by user B.

Reply:
That's not possible, because bad actors can only register with `unconfirmed_email`. For the app to move the value into the `email` column, someone would have to click the link sent to security@rubygems.org.

Reply:
Thanks @indirect for explaining, I'd forgotten about the unconfirmed_email/email behaviour, much appreciated.